starknet_py/compile/compiler.py

```diff
@@ -0,0 +1,133 @@
+import json
+import os
+from pathlib import Path
+from typing import List, NewType, Optional, Tuple, Union
+
+from starkware.starknet.services.api.contract_definition import ContractDefinition
+from starkware.cairo.lang.compiler.constants import MAIN_SCOPE, LIBS_DIR_ENVVAR
+from starkware.cairo.lang.cairo_constants import DEFAULT_PRIME
+from starkware.cairo.lang.compiler.cairo_compile import (
+    get_module_reader,
+)
+from starkware.cairo.lang.compiler.preprocessor.preprocess_codes import preprocess_codes
+from starkware.starknet.compiler.compile import assemble_starknet_contract
+from starkware.starknet.compiler.starknet_pass_manager import starknet_pass_manager
+
+CairoSourceCode = NewType("CairoSourceCode", str)
+CairoFilename = NewType("CairoFilename", str)
+StarknetCompilationSource = NewType(
+    "CairoSource", Union[CairoSourceCode, List[CairoFilename]]
+)
+
+
+class Compiler:
+    """
+    Class for compiling Cairo contracts
+    """
+
+    def __init__(
+        self,
+        compiled_contract: Optional[str] = None,
+        contract_source: Optional[List[StarknetCompilationSource]] = None,
+        search_paths: Optional[List[str]] = None,
+    ):
+        """
+        Initializes compiler.
+
+        :param compiled_contract: an already compiled contract
+        :param contract_source: a list of source files paths
+        :param search_paths: a ``list`` of paths used by starknet_compile to resolve dependencies within contracts
+        """
+        if not contract_source and not compiled_contract:
+            raise ValueError(
+                "One of compiled_contract or compilation_source is required."
+            )
+
+        self.compiled_contract = compiled_contract
+        self.contract_source = contract_source
+        self.search_paths = search_paths
+
+    def create_contract_definition(
```

Comment:
This method is currently only used in `contract.py` so maybe it could be a static method (or event just a method) on Contract?

Reply:
Whichever you are more comfortable with. I think it's weird to pass a compiled contract to the compiler, and since it creates contract definition it might live on contract, but wouldn't it bloat the code too much?